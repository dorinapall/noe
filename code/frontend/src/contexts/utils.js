export async function handleRequest(requestCreator) {
  let response;
  try {
    response = await requestCreator();
    response.error = false;
  } catch (error) {
    if (!error.response) {
      return { error: true };
    }
    response = error.response;
    setErrorFlagsOnResponse(response);
  }
  return response;
}

function setErrorFlagsOnResponse(response) {
  response.error = true;
  response.errors = response.data;
}

export function addStateToLocalStorage(state) {
  localStorage.setItem("tesztallomasState", JSON.stringify(state));
}

export function loadStateFromLocalStorage(setters) {
  setters.setAppointmentState(JSON.parse(localStorage.getItem("appointmentState")));
  setters.setSeatState(JSON.parse(localStorage.getItem("seatState")));
  setters.setSurveyState(JSON.parse(localStorage.getItem("surveyState")));
  localStorage.clear();
}
