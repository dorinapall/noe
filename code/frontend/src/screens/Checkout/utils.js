import React from "react";
import moment from "moment";

export function formatAppointmentDate(selectedTimeSlot) {
  if (!selectedTimeSlot) {
    return "";
  }

  const start = moment(selectedTimeSlot.start);
  return (
    <>
      {`${start.format("YYYY. MM. DD.")}`}
      <br />
      {`${start.format("HH:mm")}-${moment(selectedTimeSlot.end).format(
        "HH:mm"
      )}`}
    </>
  );
}

export function canDeleteSeat(seats) {
  return seats.length > 1;
}
