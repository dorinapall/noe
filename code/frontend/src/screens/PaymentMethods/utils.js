export function makePaymentUpdateRequest(appointment) {
  return {
    appointment: appointment.url,
    payment_method_type: "ON_SITE",
    total_price: appointment.total_price,
    currency: appointment.currency,
  };
}

export function getTotalPriceDisplay(appointment) {
  let total = "Ár nem elérhető!";
  if (
    appointment.total_price !== undefined &&
    appointment.currency !== undefined
  ) {
    const currency =
      appointment.currency === "HUF" ? "Ft" : appointment.currency;
    total = `${appointment.total_price} ${currency}`;
  }

  return total;
}
