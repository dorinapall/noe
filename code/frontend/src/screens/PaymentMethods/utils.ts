import { useFeatureSimplePay } from "../../featureFlags";
import { Appointment } from "../../models";

export function makePaymentUpdateRequest(appointment: Appointment, productID, billingDetailsValues, paymentMethod) {
  return {
    appointment: appointment.url,
    total_price: appointment.totalPrice,
    currency: appointment.currency,
    product_type: productID,
    ...(useFeatureSimplePay ? { payment_method: paymentMethod } : {}),
    ...billingDetailsValues,
  };
}

export function getTotalPriceDisplay(appointment: Appointment) {
  if (!appointment.currency) {
    return "Ár nem elérhető!";
  }

  let currency = appointment.currency;
  if (currency === "HUF") {
    currency = "Ft";
  }

  return `${appointment.totalPrice} ${currency}`;
}
