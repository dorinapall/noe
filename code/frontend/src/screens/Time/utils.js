import moment from "moment";

export function makeDateTimeFromDate(dateString) {
  let momentValue = moment();
  if (!!dateString) {
    momentValue = moment(`${dateString} ${moment().format("HH:mm:ss")}`);
  }

  return momentValue
    .toISOString(true) // Keep utc offset
    .replace(/\.[0-9]{3}/g, ""); // cut off milliseconds
}

export function getTimeSlotOptions(timeSlots) {
  if (timeSlots === null) {
    return [];
  }

  return timeSlots.map((slot) => ({
    value: slot.url,
    text: `${moment(slot.start).format("HH:mm")}-${moment(slot.end).format(
      "HH:mm"
    )}`,
  }));
}

export function updateFiltersWithDate(filters, newDate) {
  const newDateValue = makeDateTimeFromDate(newDate);

  // Filters for date, but expects an isostring to
  // determine the client's timezone
  filters.start_date = newDateValue;
}
