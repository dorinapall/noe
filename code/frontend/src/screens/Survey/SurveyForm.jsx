import React from "react";
import { useForm } from "react-hook-form";
import * as surveyUtils from "./utils";

import { Form, Field, NextButton, Toggle, InputGroup, Label } from "../../UI";

const toggleOptions = [
  { value: "yes", text: "Igen" },
  { value: "no", text: "Nem" },
];

export default function SurveyForm({
  submitMode,
  surveyQuestions,
  surveyAnswersForActiveSeat,
  onSubmit,
}) {
  const formData = surveyUtils.getFormData(
    submitMode,
    surveyQuestions,
    surveyAnswersForActiveSeat
  );
  const { register, handleSubmit, errors, setError } = useForm();

  return (
    <Form onSubmit={handleSubmit((values) => onSubmit(values, setError))}>
      {formData.map((field) => {
        // May be a question (create) or an answer (update)
        if (field.type === "survey-toggle") {
          return (
            <InputGroup key={field.name}>
              <Label>{field.label}</Label>
              <Toggle
                register={register}
                options={toggleOptions}
                name={field.name} // question url for creation, answer url for update
                defaultValue={field.defaultValue}
              />
            </InputGroup>
          );
        }
        return (
          <Field
            key={field.name}
            register={register}
            label={field.label}
            errors={errors}
            name={field.name} // question url for creation, answer url for update
            type={field.type}
            defaultValue={field.defaultValue}
          />
        );
      })}
      <NextButton type="submit" />
    </Form>
  );
}
