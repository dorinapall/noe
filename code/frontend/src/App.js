import React from "react";
import {
  BrowserRouter as Router,
  Switch,
  Route,
  Redirect,
} from "react-router-dom";

import "./styles/styles.css";
import { Provider as AppointmentProvider } from "./contexts/appointmentContext";
import { Provider as LocationProvider } from "./contexts/locationContext";
import { Provider as SeatProvider } from "./contexts/seatContext";
import { Provider as SurveyProvider } from "./contexts/surveyContext";
import Nav from "./components/Nav";
import Start from "./screens/Start/Start";
import AddSeat from "./screens/AddSeat";
import EmailVerification from "./screens/EmailVerification";
import VerifyEmail from "./screens/VerifyEmail";
import PaymentMethod from "./screens/PaymentMethod";
import Registration from "./screens/Registration/Registration";
import SeatDetails from "./screens/SeatDetails/SeatDetails";
import Survey from "./screens/Survey";
import Time from "./screens/Time";
import Checkout from "./screens/Checkout";
import AppointmentSuccess from "./screens/AppointmentSuccess";

export const ROUTE_START = "/start";
export const ROUTE_EMAIL_VERIFICATION = "/megerosito-email";
export const ROUTE_VERIFY_EMAIL = "/email-megerosites";
export const ROUTE_REGISTRATION = "/regisztracio";
export const ROUTE_SEAT_DETAILS = "/szemelyes-adatok";
export const ROUTE_SURVEY = "/kerdoiv";
export const ROUTE_ADD_SEAT = "/uj-szemely";
export const ROUTE_TIME = "/idopont";
export const ROUTE_PAYMENT_METHODS = "/fizetesi-mod";
export const ROUTE_CHECKOUT = "/osszegzes";
export const ROUTE_APPOINTMENT_SUCCESS = "/sikeres-regisztracio";

let DEFAULT_ROUTE = ROUTE_START;
if (process.env.NODE_ENV === "development") {
  DEFAULT_ROUTE = ROUTE_TIME;
}

function App() {
  const routes = [
    { path: ROUTE_START, component: Start },
    { path: ROUTE_EMAIL_VERIFICATION, component: EmailVerification },
    { path: ROUTE_VERIFY_EMAIL, component: VerifyEmail },
    { path: ROUTE_REGISTRATION, component: Registration },
    { path: ROUTE_SEAT_DETAILS, component: SeatDetails },
    { path: ROUTE_SURVEY, component: Survey },
    { path: ROUTE_ADD_SEAT, component: AddSeat },
    { path: ROUTE_TIME, component: Time },
    { path: ROUTE_PAYMENT_METHODS, component: PaymentMethod },
    { path: ROUTE_CHECKOUT, component: Checkout },
    { path: ROUTE_APPOINTMENT_SUCCESS, component: AppointmentSuccess },
  ];

  return (
    <div className="App">
      <Router>
        <Nav routes={routes} />
        <Switch>
          <Redirect exact from="/" to={DEFAULT_ROUTE} />
          {routes.map((route) => (
            <Route path={route.path} key={route.path}>
              <route.component></route.component>
            </Route>
          ))}
        </Switch>
      </Router>
    </div>
  );
}

function WrappedApp(props) {
  return (
    <AppointmentProvider>
      <LocationProvider>
        <SeatProvider>
          <SurveyProvider>
            <App />
          </SurveyProvider>
        </SeatProvider>
      </LocationProvider>
    </AppointmentProvider>
  );
}

export default WrappedApp;
