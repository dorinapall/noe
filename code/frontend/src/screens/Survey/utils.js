import { ROUTE_ADD_SEAT, ROUTE_CHECKOUT } from "../../App";

export const SUBMIT_MODE_CREATE = "CREATE";
export const SUBMIT_MODE_UPDATE = "UPDATE";

export function getFieldTypeFromSurveyAnswerType(question) {
  switch (question.answer_datatype) {
    case "BOOLEAN":
      return "survey-toggle";
    case "INTEGER":
      return "number";
    // including `STRING`
    default:
      return "text";
  }
}

export function getSubmitMode(surveyAnswersForActiveSeat) {
  return surveyAnswersForActiveSeat === null
    ? SUBMIT_MODE_CREATE
    : SUBMIT_MODE_UPDATE;
}

export function getFormData(
  submitMode,
  surveyQuestions,
  surveyAnswersForActiveSeat
) {
  if (submitMode === SUBMIT_MODE_CREATE) {
    return getFormDataForCreation(surveyQuestions);
  }

  return getFormDataForUpdate(surveyQuestions, surveyAnswersForActiveSeat);
}

function getFormDataForCreation(surveyQuestions) {
  return surveyQuestions.map((question, i) => {
    const fieldType = getFieldTypeFromSurveyAnswerType(question);
    return {
      label: question.question,
      name: `question-${i}`, // To create answers based on the question url
      type: fieldType,
      defaultValue: fieldType === "survey-toggle" ? "no" : "",
    };
  });
}

function getFormDataForUpdate(surveyQuestions, surveyAnswersForActiveSeat) {
  return surveyQuestions.map((question, i) => {
    const existingAnswer = surveyAnswersForActiveSeat.filter(
      (answer) => answer.question === question.url
    )[0];
    return {
      label: question.question,
      name: `answer-${i}`, // To update existing answers based on their urls
      type: getFieldTypeFromSurveyAnswerType(question),
      defaultValue: existingAnswer.answer,
    };
  });
}

export const processCreateValues = (values, activeSeat, surveyQuestions) => {
  return Object.keys(values).map((questionFieldName) => {
    const index = parseInt(questionFieldName.split("-")[1], 10);
    const question = surveyQuestions[index];
    return {
      question: question.url,
      answer: values[questionFieldName],
      seat: activeSeat.url,
    };
  });
};

export const processUpdateValues = (values, surveyAnswersForActiveSeat) => {
  return Object.keys(values).map((answerFieldName) => {
    const index = parseInt(answerFieldName.split("-")[1], 10);
    const answer = surveyAnswersForActiveSeat[index];
    return {
      url: answer.url,
      answer: values[answerFieldName],
      seat: answer.seat,
      question: answer.question,
    };
  });
};

export function getRedirectRoute(submitMode) {
  return submitMode === SUBMIT_MODE_CREATE ? ROUTE_ADD_SEAT : ROUTE_CHECKOUT;
}

export function matchQuestionErrors(errors, questions) {
  return matchErrors(errors, questions, "question-");
}

export function matchAnswerErrors(errors, existingAnswers) {
  return matchErrors(errors, existingAnswers, "answer-");
}

export function matchErrors(errors, data, prefix) {
  // Data can be list of question, or a list of existing answers
  // Matches error keys (urls) with form fields (question or answer)

  if (errors === undefined) {
    return {};
  }

  const matchedErrors = {};
  Object.keys(errors).forEach((error) => {
    const questionUrl = error;
    let matchingIndex = null;
    data.forEach((q, i) => {
      if (q.url === questionUrl) {
        matchingIndex = i;
      }
    });

    if (matchingIndex !== null) {
      matchedErrors[`${prefix}${matchingIndex}`] = errors[error];
    }
  });

  return matchedErrors;
}
