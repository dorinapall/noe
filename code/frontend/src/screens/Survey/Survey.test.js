import React from "react";
import axios from "axios";
import renderer from "react-test-renderer";
import { act } from "react-dom/test-utils";
import { mount } from "enzyme";
import {
  sendSurveyAnswers,
  updateSurveyAnswers,
} from "../../contexts/surveyContext";
import SurveyForm from "./SurveyForm";
import * as surveyUtils from "./utils";
import { ROUTE_ADD_SEAT, ROUTE_CHECKOUT, ROUTE_START } from "../../App";

jest.mock("axios");

test("SurveyForm for creation renders OK", () => {
  const tree = renderer
    .create(
      <SurveyForm
        submitMode={surveyUtils.SUBMIT_MODE_CREATE}
        surveyAnswersForActiveSeat={null}
        surveyQuestions={[
          { url: "fake-url-1" },
          { url: "fake-url-2" },
          { url: "fake-url-3" },
        ]}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("SurveyForm for update renders OK", () => {
  const surveyAnswersForActiveSeat = [
    { seat: "seat-1-url", question: "fake-url-1", url: "answer-1-url" },
    { seat: "seat-1-url", question: "fake-url-2", url: "answer-2-url" },
    { seat: "seat-1-url", question: "fake-url-3", url: "answer-3-url" },
  ];
  const surveyQuestions = [
    { url: "fake-url-1" },
    { url: "fake-url-2" },
    { url: "fake-url-3" },
  ];
  const tree = renderer
    .create(
      <SurveyForm
        submitMode={surveyUtils.SUBMIT_MODE_UPDATE}
        surveyAnswersForActiveSeat={surveyAnswersForActiveSeat}
        surveyQuestions={surveyQuestions}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Survey submission doesn't break without questions", async () => {
  axios.post.mockImplementationOnce(() => Promise.resolve({ data: [] }));
  const mockDispatch = jest.fn(() => {});
  await sendSurveyAnswers(mockDispatch)();
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "SET_NEW_SURVEY_ANSWERS",
    payload: {},
  });
});

test("Survey update submission doesn't break without questions", async () => {
  axios.post.mockImplementationOnce(() => Promise.resolve({ data: [] }));
  const mockDispatch = jest.fn(() => {});
  await updateSurveyAnswers(mockDispatch)([]);
  expect(mockDispatch).toHaveBeenCalledWith({
    type: "SET_NEW_SURVEY_ANSWERS",
    payload: {},
  });
});

test("Process survey answers for create", () => {
  const submitValues = {
    "question-0": "Hello",
    "question-1": "Yolo",
  };
  const activeSeat = { url: "http://localhost:8000/api/seats/1/" };
  const surveyQuestions = [
    { url: "http://localhost:8000/api/survey-questsions/1/" },
    { url: "http://localhost:8000/api/survey-questsions/2/" },
  ];
  const expected = [
    {
      question: "http://localhost:8000/api/survey-questsions/1/",
      answer: "Hello",
      seat: "http://localhost:8000/api/seats/1/",
    },
    {
      question: "http://localhost:8000/api/survey-questsions/2/",
      answer: "Yolo",
      seat: "http://localhost:8000/api/seats/1/",
    },
  ];
  const res = surveyUtils.processCreateValues(
    submitValues,
    activeSeat,
    surveyQuestions
  );
  expect(res).toEqual(expected);
});

test("Survey Form create submit with questions as subdomains", async () => {
  const surveyQuestions = [
    {
      url:
        "https://api.noe.rollet.app/api/survey-questsions/1231-123-123-124asdf1/",
    },
    {
      url:
        "https://api.noe.rollet.app/api/survey-questsions/1232-123-123-124asdf2/",
    },
    {
      url:
        "https://api.noe.rollet.app/api/survey-questsions/1233-123-123-124asdf3/",
    },
  ];
  const activeSeat = {
    url: "https://api.noe.rollet.app/api/seats/1/",
  };

  const mockSubmit = jest.fn();
  const wrapper = mount(
    <SurveyForm
      submitMode={surveyUtils.SUBMIT_MODE_CREATE}
      surveyQuestions={surveyQuestions}
      activeSeat={activeSeat}
      surveyAnswersForActiveSeat={null}
      onSubmit={mockSubmit}
    />
  );
  const form = wrapper.find("form");

  await act(async () => {
    await form.simulate("submit");
  });

  expect(mockSubmit.mock.calls[0][0]).toEqual({
    "question-0": "",
    "question-1": "",
    "question-2": "",
  });
});

test("Processing update answers", () => {
  const seatUrl = "https://api.noe.rollet.app/api/seats/1/";
  const surveyAnswersForActiveSeat = [
    {
      seat: seatUrl,
      question: "https://api.noe.rollet.app/api/survey-questions/1/",
      url: "https://api.noe.rollet.app/api/survey-answers/1/",
      answer: "",
    },
    {
      seat: seatUrl,
      question: "https://api.noe.rollet.app/api/survey-questions/2/",
      url: "https://api.noe.rollet.app/api/survey-answers/2/",
      answer: "",
    },
  ];
  const submitValues = {
    "answer-0": "Hello",
    "answer-1": "Yolo",
  };
  const expected = [
    {
      url: "https://api.noe.rollet.app/api/survey-answers/1/",
      question: "https://api.noe.rollet.app/api/survey-questions/1/",
      answer: "Hello",
      seat: seatUrl,
    },
    {
      url: "https://api.noe.rollet.app/api/survey-answers/2/",
      question: "https://api.noe.rollet.app/api/survey-questions/2/",
      answer: "Yolo",
      seat: seatUrl,
    },
  ];
  const res = surveyUtils.processUpdateValues(
    submitValues,
    surveyAnswersForActiveSeat
  );
  expect(res).toEqual(expected);
});

test("Redirect Route", () => {
  expect(surveyUtils.getRedirectRoute(surveyUtils.SUBMIT_MODE_CREATE)).toBe(
    ROUTE_ADD_SEAT
  );
  expect(surveyUtils.getRedirectRoute(surveyUtils.SUBMIT_MODE_UPDATE)).toBe(
    ROUTE_CHECKOUT
  );
});

test("matchQuestionErrors", () => {
  const questions = [{ url: "url1" }, { url: "url2" }, { url: "url3" }];
  const errors = { url1: "error1", url3: "error3" };

  const matchings = surveyUtils.matchQuestionErrors(errors, questions);
  const expected = { "question-0": "error1", "question-2": "error3" };
  expect(matchings).toEqual(expected);
});

test("matchQuestionErrors if no errors", () => {
  const questions = [{ url: "url1" }, { url: "url2" }, { url: "url3" }];
  const errors = {};

  const matchings = surveyUtils.matchQuestionErrors(errors, questions);
  const expected = {};
  expect(matchings).toEqual(expected);
});

test("matchQuestionErrors if errors are undefined", () => {
  const questions = [{ url: "url1" }, { url: "url2" }, { url: "url3" }];
  const errors = undefined;

  const matchings = surveyUtils.matchQuestionErrors(errors, questions);
  const expected = {};
  expect(matchings).toEqual(expected);
});

test("Test redirect route on submission", () => {
  const cases = [
    [["CREATE", undefined], ROUTE_ADD_SEAT],
    [["UPDATE", undefined], ROUTE_CHECKOUT],
    [["CREATE", ROUTE_START], ROUTE_START],
    [["CREATE", ROUTE_CHECKOUT], ROUTE_CHECKOUT],
    [["UPDATE", ROUTE_START], ROUTE_START],
    [["UPDATE", ROUTE_CHECKOUT], ROUTE_CHECKOUT],
  ];
  cases.forEach((testCase) => {
    expect(surveyUtils.getRedirectRoute(...testCase[0])).toBe(testCase[1]);
  });
});
