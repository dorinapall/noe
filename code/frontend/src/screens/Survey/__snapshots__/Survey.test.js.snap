// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SurveyForm for creation renders OK 1`] = `
<form
  className="Form"
  onSubmit={[Function]}
>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      defaultValue=""
      name="question-0"
      type="text"
    />
  </div>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      defaultValue=""
      name="question-1"
      type="text"
    />
  </div>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      defaultValue=""
      name="question-2"
      type="text"
    />
  </div>
  <button
    className="Button NextButton"
    type="submit"
  >
    Tovább
  </button>
</form>
`;

exports[`SurveyForm for update renders OK 1`] = `
<form
  className="Form"
  onSubmit={[Function]}
>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      name="answer-0"
      type="text"
    />
  </div>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      name="answer-1"
      type="text"
    />
  </div>
  <div
    className="InputGroup"
  >
    <input
      className="Input"
      name="answer-2"
      type="text"
    />
  </div>
  <button
    className="Button NextButton"
    type="submit"
  >
    Tovább
  </button>
</form>
`;
